### PR TITLE
Use SD sw_pll for digital Rx on XS3 targets

### PR DESCRIPTION
**Pull request by ed-xmos**
https://xmosjira.atlassian.net/jira/software/c/projects/UA/boards/54?assignee=557058%3A17873fbc-49d6-4971-8efb-a0de95f426a1&selectedIssue=UA-158

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -13,6 +13,12 @@ HEAD
   * RESOLVED:  Repeated old S/PDIF and ADAT samples when entering underflow state
   * CHANGED:   QUAD_SPI_FLASH replaced by XUA_QUAD_SPI_FLASH (default: 1)
   * CHANGED:   UserBufferManagementInit() now takes a sample rate parameter
+  * CHANGED:   xcore-ai targets use sigma-delta software PLL for clock recovery of
```
Comment (xross) [inline CHANGELOG.rst:16]: xcore.ai

File: lib_xua/doc/rst/arch.rst [hunk 2]
```diff
@@ -54,7 +54,9 @@ In addition :ref:`usb_audio_optional_components` shows optional components that
  * - Clockgen
    - Drives an external frequency generator (PLL) and manages
      changes between internal clocks and external clocks arising
-     from digital input.
+     from digital input. On xCORE-AI Clockgen may also work in
```
Comment (xross) [inline lib_xua/doc/rst/arch.rst:57]: xcore.ai

File: lib_xua/doc/rst/feat_spdif_rx.rst [hunk 3]
```diff
@@ -29,10 +29,11 @@ The S/PDIF receiver should be called on the appropriate tile::
 With the steps above an S/PDIF stream can be captured by the xCORE. To be functionally useful the audio
 master clock must be able to synchronise to this external digital stream. Additionally, the host can be 
 notified regarding changes in the validity of this stream, it's frequency etc. To synchronise to external 
-streams the codebase assumes the use of an external Cirrus Logic CS2100 device.
+streams the codebase assumes the use of an external Cirrus Logic CS2100 device or lib_sw_pll on xCORE-AI designs.
```
Comment (xross) [inline lib_xua/doc/rst/feat_spdif_rx.rst:32]: xcore.ai

File: lib_xua/doc/rst/hw.rst [hunk 4]
```diff
@@ -52,11 +52,11 @@ Three methods of generating an audio master clock are provided on the board:
 
     * A Skyworks Si5351B PLL device. The Si5351 is an I2C configurable clock generator that is ideally suited for replacing crystals, crystal oscillators, VCXOs, phase-locked loops (PLLs), and fanout buffers.
 
-    * xCORE.ai devices are equipped with a secondary (or 'application') PLL which can be used to generate audio clocks
+    * xCORE.ai devices are equipped with a secondary (or 'application') PLL which can be used to generate fixed audio clocks or recover external clocks using lib_sw_pll.
 
 Selection between these methods is done via writing to bits 6 and 7 of PORT 8D on tile[0]. 
 
-Either the locally generated clock (from the PL611) or the recovered low jitter clock (from the CS2100) may be selected to clock the audio stages; the xCORE-200, the ADC/DAC and Digital output stages. Selection is controlled via an additional I/O, bit 5 of PORT 8C, see :ref:`hw_316_ctrlport`.
+Either the locally generated clock (from the PL611) or the recovered low jitter clock (from the CS2100) may be selected to clock the audio stages; the xCORE-AI, the ADC/DAC and Digital output stages. Selection is controlled via an additional I/O, bit 5 of PORT 8C, see :ref:`hw_316_ctrlport`.
```
Comment (xross) [inline lib_xua/doc/rst/hw.rst:59]: xcore.ai

File: lib_xua/doc/rst/sw_clocking.rst [hunk 5]
```diff
@@ -15,32 +15,35 @@ the xCORE.
 
 Using an external PLL/Clock Multiplier allows an Asynchronous mode design to lock to an external
 clock source from a digital stream (e.g. S/PDIF or ADAT input).  The codebase supports the Cirrus
-Logic CS2100 device for this purpose. Other devices may be supported via code modification.
+Logic CS2100 device or use of lib_sw_pll (xCORE-AI only) for this purpose. Other devices may be
```
Comment (xross) [inline lib_xua/doc/rst/sw_clocking.rst:18]: xcore.ai

**Review by xross — CHANGES_REQUESTED** [review]
Looks good to me, just small doc updates I highlighted. 

The wrapper, im not sure exactly why this doesn't live in lib_sw_pll but it's not a biggy for the moment. 

Looks like a fantastic addition to the codebase.

**Comment by ed-xmos** [comment]
> Looks good to me, just small doc updates I highlighted.
> 
> The wrapper, im not sure exactly why this doesn't live in lib_sw_pll but it's not a biggy for the moment.
> 
> Looks like a fantastic addition to the codebase.

Yes I was wondering about the phase frequency detector logic living in sw_pll. It's subtly different from the stock one and also the SD is wrapped up in a loop with a select and other resource stuff and wanted to keep sw_pll generic. There's an SD example in sw_pll (using C rather than XC) but it differs in that the PI controller runs in client rather than the server as in usb_audio. In UA we need to keep the blocking time in clock gen very low due to 192kHz cases.. By keeping the lib generic you are not enforcing usage. I can see arguments both ways really.

**Review by xross — DISMISSED** [review]

File: lib_xua/src/core/clocking/clockgen.xc [hunk 6]
```diff
@@ -107,6 +107,7 @@ static inline void setClockValidity(chanend c_interruptControl, int clkIndex, in
     {
         clockValid[clkIndex] = valid;
         outInterrupt(c_interruptControl, clockId[clkIndex]);
+        printstr("clockValid=");printintln(valid);
```
Comment (xross) [inline lib_xua/src/core/clocking/clockgen.xc:110]: Accidental debug print left in?

**Review by xross — APPROVED** [review]

**Pull request merged**